### PR TITLE
Feature multi page api requests

### PR DESCRIPTION
**Pull request by a-szotyori**
Integrated new api requests for creating partial and composite documents to analyze single and multi-page documents.

Created the `NetworkRequestsManager` class to deal with requests between the UI and the networking library. Uses `CompletableFuture`s to chain handlers and tasks. It takes care of:
* uploading and analyzing documents only once,
* caching documents,
* waiting for partial uploads to finish before creating a composite document and getting the extractions,
* cancelling requests (currently only polling can be really cancelled).

Updated the `GiniVisionNetworkService` to work with `CancellationToken`s for cancelling requests.

Updated the `GiniVisionDefaultNetworkService` and the `GiniVisionDefaultNetworkApi` to use the latest Gini API SDK beta to work with partial documents and composite documents.

### How to test
Using Nico's local service.

### Ticket 
Issue #158

### CODE REVIEW

File: ginivision-network/src/main/java/net/gini/android/vision/network/GiniVisionDefaultNetworkService.java [hunk 1]
```diff
@@ -24,56 +38,221 @@
...
+            }
+            documentRotationMap.put(document, entry.getValue());
+        }
+        mAnalyzedApiDocument = null;
+        final AtomicBoolean isCancelled = new AtomicBoolean();
+        final AtomicReference<net.gini.android.models.Document> compositeDocument =
+                new AtomicReference<>();
+        mGiniApi.getDocumentTaskManager().createCompositeDocument(documentRotationMap, null)
+                .onSuccessTask(
+                        new Continuation<net.gini.android.models.Document, Task<net.gini.android.models.Document>>() {
```
[the remark this answers is not in the document]
Comment (a-szotyori) [reply ginivision-network/src/main/java/net/gini/android/vision/network/GiniVisionDefaultNetworkService.java]: I would rather use lambdas. I will see what I can extract into methods, although I can't promise anything. Continuations are classes and I have to see which instructions I can move into methods to reuse, if there are any.

File: ginivision/src/main/java/net/gini/android/vision/analysis/AnalysisFragmentImpl.java [hunk 2]
```diff
@@ -389,43 +433,52 @@ private void doAnalyzeDocument() {
...
+                if (mDocument instanceof GiniVisionMultiPageDocument) {
+                    multiPageDocument = (GiniVisionMultiPageDocument) mDocument;
+                } else {
+                    multiPageDocument = DocumentFactory.newMultiPageDocument(mDocument);
+                }
+                for (final Object document : multiPageDocument.getDocuments()) {
+                    final GiniVisionDocument giniVisionDocument = (GiniVisionDocument) document;
+                    networkRequestsManager.upload(activity, giniVisionDocument);
+                }
+                networkRequestsManager.analyze(multiPageDocument)
```
[the remark this answers is not in the document]
Comment (a-szotyori) [reply ginivision/src/main/java/net/gini/android/vision/analysis/AnalysisFragmentImpl.java]: That's one of the reasons behind the `NetworkRequestsManager`. Our Fragments are already too large and you are right, that a refactoring is needed. I can imagine using a standard MVP approach. Unfortunately due to time constraints and different priorities I cannot tell when or if we'll refactor it.

File: ginivision/src/main/java/net/gini/android/vision/camera/CameraFragmentImpl.java [hunk 3]
```diff
@@ -569,22 +574,65 @@ void onStop() {
...
+                    .internal().getNetworkRequestsManager();
+            if (networkRequestsManager != null) {
+                networkRequestsManager.cancel(mMultiPageDocument);
+                networkRequestsManager.delete(mMultiPageDocument)
+                        .handle(new CompletableFuture.BiFun<NetworkRequestResult<GiniVisionDocument>, Throwable, Void>() {
+                            @Override
+                            public Void apply(
+                                    final NetworkRequestResult<GiniVisionDocument> requestResult,
+                                    final Throwable throwable) {
+                                for (final Object document : mMultiPageDocument.getDocuments()) {
```
[the remark this answers is not in the document]
Comment (a-szotyori) [reply ginivision/src/main/java/net/gini/android/vision/camera/CameraFragmentImpl.java]: The Fragment is not the view it's a view controller so it having business logic is acceptable. Still something like MVP would be beneficial, but as mentioned before there's no time for a big refactoring.

I already extracted a lot of logic into the `NetworkRequestsManager`, but each screen (Camera, Review and Analysis) has to control when and which requests are needed.

File: ginivision/src/main/java/net/gini/android/vision/internal/network/NetworkRequestsManager.java [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+                        }
+                        return true;
+                    }
+                })
+                .thenCompose(
+                        new CompletableFuture.Fun<Boolean, CompletableFuture<NetworkRequestResult<GiniVisionDocument>>>() {
+                            @Override
+                            public CompletableFuture<NetworkRequestResult<GiniVisionDocument>> apply(
+                                    final Boolean success) {
+                                final CompletableFuture<NetworkRequestResult<GiniVisionDocument>>
```
[the remark this answers is not in the document]
Comment (a-szotyori) [reply ginivision/src/main/java/net/gini/android/vision/internal/network/NetworkRequestsManager.java]: 👍 I'll see what I can do to improve readability.

File: ginivision/src/main/java/net/gini/android/vision/analysis/AnalysisFragmentImpl.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kikettas) [reply ginivision/src/main/java/net/gini/android/vision/analysis/AnalysisFragmentImpl.java]: Not for this release but for the next one then 😉. Create an issue for that so we won't forget about it.

File: ginivision-network/src/main/java/net/gini/android/vision/network/GiniVisionDefaultNetworkService.java [hunk 5]
```diff
@@ -24,56 +38,221 @@
...
     @Override
-    public void analyze(@NonNull final Document document,
-            @NonNull final GiniVisionNetworkCallback<AnalysisResult, Error> callback) {
-        mSingleDocumentAnalyzer.analyzeDocument(document,
-                new DocumentAnalyzer.Listener() {
+    public CancellationToken upload(@NonNull final Document document,
+            @NonNull final GiniVisionNetworkCallback<Result, Error> callback) {
+        LOG.debug("Upload document {}", document.getId());
+        if (document.getData() == null) {
+            final Error error = new Error("Document has no data. Did you forget to load it?");
```
[the remark this answers is not in the document]
Comment (a-szotyori) [reply ginivision-network/src/main/java/net/gini/android/vision/network/GiniVisionDefaultNetworkService.java]: The error messages are not intended for users. That is also the reason why the messages are not localized. I will make it clear in the documentation later.

File: ginivision-network/src/main/java/net/gini/android/vision/network/GiniVisionDefaultNetworkService.java [hunk 6]
```diff
@@ -24,56 +38,221 @@
...
     @Override
-    public void cancel() {
-        mSingleDocumentAnalyzer.cancelAnalysis();
+    public CancellationToken analyze(
+            @NonNull final LinkedHashMap<String, Integer> documentIdRotationMap,
+            @NonNull final GiniVisionNetworkCallback<AnalysisResult, Error> callback) {
+        LOG.debug("Analyze documents {}", documentIdRotationMap);
+        final LinkedHashMap<net.gini.android.models.Document, Integer> documentRotationMap =
+                new LinkedHashMap<>();
+        for (final Map.Entry<String, Integer> entry : documentIdRotationMap.entrySet()) {
```
[the remark this answers is not in the document]
Comment (a-szotyori) [reply ginivision-network/src/main/java/net/gini/android/vision/network/GiniVisionDefaultNetworkService.java]: I extracted creating the map. Didn't find a way to make the method shorter. Even though it's long I think it's easier to understand this way than extract the continuations into methods or dedicated classes.

File: ginivision/src/main/java/net/gini/android/vision/analysis/AnalysisFragmentImpl.java [hunk 7]
```diff
@@ -190,19 +213,40 @@ public View onCreateView(final LayoutInflater inflater, final ViewGroup containe
...
+                    .internal().getNetworkRequestsManager();
+            if (networkRequestsManager != null) {
+                networkRequestsManager.cancel(mDocument);
+                networkRequestsManager.delete(mDocument)
+                        .handle(new CompletableFuture.BiFun<NetworkRequestResult<GiniVisionDocument>, Throwable, Void>() {
+                            @Override
+                            public Void apply(
+                                    final NetworkRequestResult<GiniVisionDocument> requestResult,
+                                    final Throwable throwable) {
+                                if (mDocument instanceof GiniVisionMultiPageDocument) {
```
[the remark this answers is not in the document]
Comment (kikettas) [reply ginivision/src/main/java/net/gini/android/vision/analysis/AnalysisFragmentImpl.java]: Ok. Don't forget about this!

**Comment by a-szotyori** [comment]
@kikettas Updated document deletion when users leaves a screen:
* Camera Screen deletes multi-page document, if it exists
* Single Page Review Screen always deletes its document
* Multi-Page Review Screen deletes only "open with" documents. Any other document will be deleted by the Camera Screen.
* Analysis Screen deletes only PDF documents. Any other document will be deleted by the Review or Camera Screen.

**Pull request merged**